sockeye/scoring.py

```diff
@@ -0,0 +1,481 @@
+                 score: float) -> None:
+        self.sentence_id = sentence_id
+        self.source_tokens = source_tokens
+        self.target_tokens = target_tokens
+        self.score = score
+
+    def __str__(self):
+        """
+        Returns a string representation of a scoring output.
+        """
+        return 'ScoringOutput(%d, %s, %s, %f)' % (self.sentence_id, self.source_tokens, self.target_tokens, self.score)
+
+
+ScoredBatch = List[Tuple[int, float]]
+ScoredSamples = List[ScoredBatch]
+
+
+class Scorer:
+    """
+    Scorer uses one or several models to score pairs of input strings.
+
+    :param batch_size: Batch size.
+    :param context: The context(s) that MXNet will be run in (GPU(s)/CPU).
+    :param ensemble_mode: Ensemble mode: linear combination of scores.
+    :param no_bucketing: If False bucketing will be used, if True the
+    computation graph will always be unrolled to the full length.
+    :param normalize: If True, normalize scores by the length of the target
+    sequence.
+    """
+    def __init__(self,
+                 batch_size: int,
+                 context: List[mx.context.Context],
+                 ensemble_mode: str,
+                 no_bucketing: bool = False,
+                 normalize: Optional[bool] = True) -> None:
+
+        self.batch_size = batch_size
+        self.context = context
+        self.no_bucketing = no_bucketing
+        self.normalize = normalize
+        self.interpolation_func = self._get_interpolation_func(ensemble_mode)
+
+    @staticmethod
+    def _get_interpolation_func(ensemble_mode):
+        if ensemble_mode == 'linear':
+            return Scorer._linear_interpolation
+        else:
+            raise ValueError("unknown interpolation type")
+
+    @staticmethod
+    def _linear_interpolation(predictions):
+        return np.mean(predictions, axis=0)
+
+    def _score_batch(self,
+                     model: ScoringModel,
+                     batch: mx.io.DataBatch,
+                     batch_index: Tuple[int, int],
+                     mapid: np.ndarray
+                     ) -> ScoredBatch:
+        """
+        Scores a batch of examples, returns a list of tuples (sentence_id, score).
+        :param model: The model used for scoring, an instance of ScoringModel.
+        :param batch: A batch of data that is forwarded through the model.
+        :param batch_index: An identifier for the specific batch.
+        :param mapid: Nested dictionary mapping  positions in buckets to the
+        original ordering in the input.
+        :return: A ScoredBatch which is a list of tuples (sentence_id, score).
+        """
+        scored_batch = []
+
+        model.module.forward(batch, is_train=False)
+        outputs = model.module.get_outputs()
+
+        # split output array into probs per batch
+        sample_length = int(len(outputs[0]) / self.batch_size)
+
+        probs = mx.nd.array(outputs[0], ctx=self.context)  # shape is (t_len*batch_size, t_vocab)
```

Comment:
isnt `outputs[0]` already an ndarray?

Reply:
also, the comment is wrong I think, the shape should be `(batch_size * t_len, t_vocab)`
(a tensor of `(batch, len, vocab)` reshaped to `(batch*len, vocab)` which is a different ordering of data)